pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -595,8 +595,26 @@ func (cb *ClusterBuilder) buildLocalityLbEndpoints(proxyView model.ProxyView, se
 				Value: instance.Endpoint.GetLoadBalancingWeight(),
 			},
 		}
+
+		labels := instance.Endpoint.Labels
+		ns := instance.Endpoint.Namespace
+		if features.CanonicalServiceForMeshExternalServiceEntry && service.MeshExternal {
+			ns = service.Attributes.Namespace
+			svcLabels := service.Attributes.Labels
+			if _, ok := svcLabels[model.IstioCanonicalServiceLabelName]; ok {
+				labels = map[string]string{
```

Comment:
A bit confused that we have this and ensureCanonicalServiceLabels? Do we need both?

Reply:
_need_ is probably a bit strong. the `cluster_builder.go` bits here take from the service's labels that are built in the other bit (conversion). we _could_ have it just do the logic here, but it wasn't clear to me if the "existence" of those labels might be of use elsewhere.

Reply:
Oh this is Service vs instance. got it